source/connection-monitoring-and-pooling/connection-monitoring-and-pooling.md

```diff
@@ -576,7 +579,122 @@ other threads from checking out [Connections](#connection) while establishing a
+    timeout, or if the connection exceeds the 3 second pending-response window, the driver MUST close the connection.
+6. **Clear pending state on success**: If the pending response is fully drained and successfully discarded, and the
+    connection remains healthy, the pending state may be cleared and the connection MAY be returned to the pool for
+    reuse.
+
+```mermaid
+sequenceDiagram
+    participant Driver
+    participant Pool
+    participant Conn as Connection (*)
+    participant Server
+
+    Driver->>Pool: Checkout Connection (*)
+    Pool->>Driver: Return connection (*)
+    Driver->>Conn: Send operation (1) (CSOT enabled, maxTimeMS > 0, exhaustAllowed = false)
+    Conn->>Server: Send command
+    Server-->>Conn: (No response, socket times out)
+
+    Conn->>Conn: Transition connection to "pending response" state, record current time
+    Conn-->>Driver: Error
+
+    Driver->>Pool: Checkout Connection (*)
+    Pool->>Driver: Return connection (*)
+    Driver->>Conn: Send operation (2)
+
+    Conn->>Conn: Attempt to drain pending response from operation (1)
+    Conn->>Conn: Update pending read timestamp if bytes read
+    alt Timeout window exceeded or non-timeout error
+        Conn->>Conn: Close connection
+        Conn-->>Driver: Error
+    else Timeout window not exceeded
+       alt Error
+            Conn->>Conn: Clear pending response state
+            Conn->>Conn: Reset to current time
+            Conn->>Pool: Check connection back into pool
+            Conn-->>Driver: Error
+        else No error
+            Conn->>Conn: Clear pending response state
+            Conn->>Driver: Return connection to execute operation
+        end
+    end
+```
+
+```python
+PENDING_RESPONSE_TIMEOUT_MS = 3000  # static timeout
+
+def await_pending_response(timeout, conn):
+    # Note: conn.pending_start is initialized after the original socket timeout
+    # and not in this function since the connection will sit in the pool for some
+    # non-deterministic amount of time after the socket timeout.
+
+    remaining_time = (conn.pending_start + PENDING_RESPONSE_TIMEOUT_MS) - current_time()
+    if remaining_time <= 0:
+        # Use the smallest timeout (or enable non-blocking read).
+        remaining_time = 0.001
+
+    if timeout is None:
+        timeout = min(remaining_time, conn.socket_timeout_ms)
+    else:
+        timeout = min(remaining_time, timeout)
+
+    data, error = execute_pending_response(timeout, conn)
+    end_time = current_time()
+
+    if error is not None and error is not timeout:
+        close_connection(conn)
+        raise
+
+    if len(data) > 0:
+        # Refresh the remaining time upon a successful read
+        conn.pending_start = end_time
+
+    # Check if the remaining time has been exceeded
+    if end_time - conn.pending_start >= PENDING_RESPONSE_TIMEOUT_MS:
+        close_connection(conn)
+
+    if error is not None:
+        raise error
```

Comment:
The pseudocode that existed before the current PR used `throw`, for example, `throw PoolClosedError`. The new pseudocode should continue using that, instead of coming up with new syntax.